doc/transformations.rst

```diff
@@ -180,6 +180,50 @@ transformed version of that kernel exists and does not match that
 created by the current transformation then PSyclone will raise an
 exception.
 
+Rules
++++++
+
+Kernels that are to be transformed are subject to additional
+restrictions. These rules are intended to make kernel transformations
+as robust as possible, in particular by limiting the amount of
+code that must be parsed by PSyclone (via fparser). The rules are
+as follows:
+
+1) Any variable accessed by a kernel must either be explicitly declared
+   or named in the ``only`` clause of a module ``use`` statement within
+   the scope of the subroutine containing the kernel implementation;
+
+2) The full Fortran source of a kernel must be available to PSyclone.
+   This includes the source of any modules from which it accesses
+   either routines or data.
```

Comment:
I am not convinced with "The full Fortran source" as in the case of routines there are several Fortran mechanisms to soften the constraint: interfaces, public keyword (?), c_bindings.

An alternative would be to link 2) to the previous point and ask for the declarations of the named instances in the `use` statements to be available to (or discoverable by) PSyclone.

Reply:
Thanks @sergisiso. I've had a think about this and the problem is that we need to be sure that, when transforming kernels for either OpenACC or OpenCL, we've captured all use-dependencies in the call-tree of a kernel. However, having written that I can see that this is only a condition for those two cases - if we were just manipulating the form of a kernel routine then perhaps just having the interface information available would be sufficient. Given that OpenACC and OpenCL are our immediate targets, I think keeping this restriction as it is is OK. We can always relax it in future if/when we are doing other sorts of kernel transformation and it becomes a problem.